### PR TITLE
Feat: Initial Kad node setup

### PR DESCRIPTION
**Pull request by ryardley**
Closes: #177 
- [x] Hook up Kademlia for peer discovery
- [x] Test simple networking without mdns
- [x] Plumbed peers list from config
- [x] Rename `EnclaveRouter` to `NetworkPeer`
- [x] Refactor libp2p code to be simpler - perhaps a function or a separate builder 

## Future PR
- Auto Deploy Nodes to Bootstrap the network
- CLI to populate config with nodes from centralized (maybe github?) list then moving on to ipns

## Description

This is the initial node setup to use Kademlia in a gossipsub config.

## What's Changed

- I've removed the rendevous server
- I've added connection limits, updated the gossipsub config and services needed for Kademlia

## Notes

- Ideally you will have some bootstrap nodes that all nodes can always connect to (at least in the inchoate stages of the network)
- This needs to be tested in a deployed devnet

---

This branch was created with the awesome help of @maschad! 🙏🏾 

## Release Notes

- **New Features**
  - Introduced a new networking library named "net" with improved functionalities for peer-to-peer communication.
  - Added integration tests for the new networking features, including a test harness for nodes communicating via gossipsub messages.

- **Bug Fixes**
  - Updated dependency management to reflect changes from "p2p" to "net," ensuring compatibility with the new library structure.

- **Documentation**
  - Added instructions for running integration tests in the README file.

- **Chores**
  - Created Docker and Docker Compose configurations for testing environments.
  - Introduced new scripts for managing testing processes and enhancing error handling.

### CODE REVIEW

**Comment by ryardley** [comment]
Gonna try and use configuration to enable optional mdns get the test to pass and then see if I can get the changes to work in a docker networking test.

**Comment by ryardley** [comment]
Using config ended up being overly complex so resorted to adding mdns to the node behavior. For the test we need quic to be able to pick it's own random port as we don't have any network setup with specific ip addresses yet so I commented out the attempt to setup a cannonical quick addr.

File: packages/ciphernode/Cargo.toml [hunk 1]
```diff
@@ -54,18 +54,13 @@ sha2 = "0.10.8"
...
-  "noise",
   "ping",
-  "rendezvous",
-  "tcp",
   "tokio",
-  "yamux",
-  "mdns",
   "gossipsub",
   "quic",
 ] }
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/Cargo.toml:65]: If this were the case it wouldn't compile?

File: packages/ciphernode/enclave_node/src/aggregator.rs [hunk 2]
```diff
@@ -82,7 +82,8 @@ pub async fn setup_aggregator(
         .build()
         .await?;
 
-    let (_, join_handle, peer_id) = P2p::spawn_libp2p(bus.clone()).expect("Failed to setup libp2p");
+    let (_, join_handle, peer_id) =
+        NetworkRelay::setup_with_peer(bus.clone(), config.peers()).expect("Failed to setup libp2p");
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/enclave_node/src/aggregator.rs]: Resolved

File: packages/ciphernode/enclave_node/src/ciphernode.rs [hunk 3]
```diff
@@ -71,7 +71,8 @@ pub async fn setup_ciphernode(
         .build()
         .await?;
 
-    let (_, join_handle, peer_id) = P2p::spawn_libp2p(bus.clone()).expect("Failed to setup libp2p");
+    let (_, join_handle, peer_id) =
+        NetworkRelay::setup_with_peer(bus.clone(), config.peers()).expect("Failed to setup libp2p");
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/enclave_node/src/ciphernode.rs:75]: Resolved

File: packages/ciphernode/net/tests/entrypoint.sh [hunk 4]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+set -e
+
+if [ "${BLOCK_MDNS:-false}" = "true" ]; then
+    iptables -A INPUT -p udp --dport 5353 -j DROP
+    iptables -A OUTPUT -p udp --dport 5353 -j DROP
+    iptables -L | grep DROP
+fi
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/net/tests/entrypoint.sh:8]: this is test code not appropriate

File: packages/ciphernode/net/tests/run.sh [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+echo ""
+echo "TEST 1: Using MDNS with separate IP addresses"
+echo ""
+docker compose up --build --abort-on-container-exit
+
+echo ""
+echo "TEST 2: Blocking MDNS traffic for each service"
+echo ""
+echo "Note this should display some libp2p_mdns::behaviour::iface errors in output"
+echo ""
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/net/tests/run.sh:14]: this is test code not appropriate

File: packages/ciphernode/net/tests/run.sh [hunk 6]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+
+set -e
+
+echo ""
+echo "TEST 1: Using MDNS with separate IP addresses"
+echo ""
+docker compose up --build --abort-on-container-exit
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/net/tests/run.sh:8]: not appropriate

File: packages/ciphernode/net/tests/docker-compose.yaml [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    networks:
+      app_net:
+        ipv4_address: 172.16.238.10
+    command: cargo run --bin p2p_test alice
+    environment:
+      QUIC_PORT: 9091
+      DIAL_TO: "/ip4/172.16.238.12/udp/9091/quic-v1"
+      BLOCK_MDNS: ${BLOCK_MDNS:-false}
+    entrypoint: ["/app/entrypoint.sh"]
+    privileged: true
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/net/tests/docker-compose.yaml:15]: resolved

File: packages/ciphernode/net/src/network_relay.rs [hunk 8]
```diff
@@ -34,51 +36,47 @@ impl P2p {
...
+    ) -> Result<(Addr<Self>, tokio::task::JoinHandle<Result<()>>, String), Box<dyn Error>> {
         let keypair = libp2p::identity::Keypair::generate_ed25519();
-        libp2p.with_identity(&keypair);
-        libp2p.connect_swarm("mdns".to_string())?;
-        libp2p.join_topic("enclave-keygen-01")?;
-
-        let p2p_addr = Self::spawn_and_listen(bus, tx, rx);
-        let handle = tokio::spawn(async move { libp2p.start().await.unwrap() });
+        let mut peer = NetworkPeer::new(&keypair, peers, None, "tmp-enclave-gossip-topic")?;
+        let rx = peer.rx().ok_or(anyhow!("Peer rx already taken"))?;
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/net/src/network_relay.rs]: Not really sure as to what validation here might look like and peers are converted to multiaddrs so I assume they are validated in terms of format there. Duplicates are not really an issue as connection are deduplicated.

**Comment by ryardley** [comment]
In terms of tidying up and refactoring:

- Renamed to `NetworkRelay` which forwards events over the lower level `NetworkPeer` 
- Moved the `p2p` crate to `net` in case we need to do any other networking operations there.
- Removed the optionality over the swarm which was adding complexity.
- Removed the confusing way to handle channels in the constructor of the NetworkPeer.

**Comment by hmzakhalid** [comment]
Everything Looks good and works well, Just a small little update to the test dockerfile
From the PR #187 

> Uses a multistage build to reduce the Docker image size from 3.5 GB to 100 MB 
I also noticed that it was creating three separate images for the same build because each service in the Docker Compose file had its own build context. Instead, only one image should be built and shared across all the services, with each service simply instantiating its own container from that shared image.
 Building the image first and then referencing it in the services resolves this issue.

**Review by hmzakhalid — APPROVED** [review]
lgtm

**Pull request merged**